Source/Core/Common/WorkQueueThread.h

```diff
@@ -0,0 +1,86 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <functional>
+#include <queue>
+#include <thread>
+
+#include "Common/Event.h"
+#include "Common/Flag.h"
+
+// A thread that executes the given function for every item placed into its queue.
+
+namespace Common
+{
+template <typename T>
+class WorkQueueThread
+{
+public:
+  WorkQueueThread() = default;
+  WorkQueueThread(std::function<void(T)> function) { Reset(std::move(function)); }
+  ~WorkQueueThread() { Shutdown(); }
+  void Reset(std::function<void(T)> function)
+  {
+    Shutdown();
+    m_shutdown.Clear();
+    m_function = std::move(function);
+    m_thread = std::thread([this] { ThreadLoop(); });
+  }
+
+  template <typename... Args>
+  void EmplaceItem(Args&&... args)
+  {
+    {
+      std::unique_lock<std::mutex> lg(m_lock);
+      m_items.emplace(std::move(args)...);
```

Comment:
The better way to pass arguments along here would be to use `std::forward`, otherwise perfect forwarding can't occur.